tools/node-modules

```diff
@@ -0,0 +1,223 @@
+#!/bin/sh
+
+set -eu
+
+GITHUB_BASE="${GITHUB_BASE:-cockpit-project/cockpit}"
+GITHUB_REPOSITORY="${GITHUB_BASE%/*}/node-cache"
+CLONE_REMOTE="https://github.com/${GITHUB_REPOSITORY}"
+PUSH_REMOTE="git@github.com:${GITHUB_REPOSITORY}"
+
+CACHE_DIR="${XDG_CACHE_HOME-${HOME}/.cache}/cockpit-dev/${GITHUB_REPOSITORY}.git"
+
+message() {
+    [ "${V-}" != 0 ] || printf "  %-8s %s\n" "$1" "$2"
+}
+
+get_index_gitlink() {
+    if ! git ls-files -s "$1" | egrep -o '\<[[:xdigit:]]{40}\>'; then
+        echo "*** couldn't read gitlink for file $1 from the index" >&2
+        exit 1
+    fi
+}
+
+cmd_remove() {
+    # if we did this for ourselves the rm is enough, but it might be the case
+    # that someone actually used git-submodule to fetch this, so clean up after
+    # that as well.  NB: deinit nicely recreates the empty directory for us.
+    message REMOVE node_modules
+    rm -rf node_modules
+    git submodule deinit node_modules
+    rm -rf "$(git rev-parse --absolute-git-dir)/modules/node_modules"
+}
+
+cmd_checkout() {
+    # we check out the node_modules corresponding to the gitlink in the index
+    local sha="$(get_index_gitlink node_modules)"
+    local tag="sha-${sha}"
+
+    if [ ! -d "${CACHE_DIR}" ]; then
+        message INIT "${CACHE_DIR}"
+        mkdir -p "${CACHE_DIR}"
+        git --git-dir "${CACHE_DIR}" init ${quiet}
+        git --git-dir "${CACHE_DIR}" remote add origin "${CLONE_REMOTE}"
+    fi
+
+    # fetch the tag if we don't already have it
+    if ! git --git-dir "${CACHE_DIR}" rev-parse --verify --quiet "${tag}" >/dev/null; then
+        message FETCH "node_modules  [sha: ${sha}]"
+        # fetch by sha to prevent us from downloading something we don't want
+        # the tag should come automatically with it
+        git --git-dir "${CACHE_DIR}" fetch ${quiet} --tags origin "${sha}"
```

Comment:
`--tags` fetches *all* tags, not just the one you want. I run into the same trouble in #15748, which is why I started this with `FETCH_HEAD` (which is of course a bit racy)

Reply:
As always, [stackoverflow to the rescue](https://stackoverflow.com/questions/45338495/fetch-a-single-tag-from-remote-repository): `git fetch origin tag "$tag" --no-tags`